src/main/java/org/opensearch/neuralsearch/query/SparseEncodingQueryBuilder.java

```diff
@@ -0,0 +1,266 @@
+                    sparseEncodingQueryBuilder.queryText(parser.text());
+                } else if (MODEL_ID_FIELD.match(currentFieldName, parser.getDeprecationHandler())) {
+                    sparseEncodingQueryBuilder.modelId(parser.text());
+                } else {
+                    throw new ParsingException(
+                        parser.getTokenLocation(),
+                        "[" + NAME + "] query does not support [" + currentFieldName + "]"
+                    );
+                }
+            } else {
+                throw new ParsingException(
+                    parser.getTokenLocation(),
+                    "[" + NAME + "] unknown token [" + token + "] after [" + currentFieldName + "]"
+                );
+            }
+        }
+    }
+
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
+        // We need to inference the sentence to get the queryTokens. The logic is similar to NeuralQueryBuilder
+        // If the inference is finished, then rewrite to self and call doToQuery, otherwise, continue doRewrite
+        if (null != queryTokensSupplier) {
+            return this;
+        }
+
+        validateForRewrite(queryText, modelId);
+        SetOnce<Map<String, Float>> queryTokensSetOnce = new SetOnce<>();
+        queryRewriteContext.registerAsyncAction(
+            ((client, actionListener) -> ML_CLIENT.inferenceSentencesWithMapResult(
+                modelId(),
+                List.of(queryText),
+                ActionListener.wrap(mapResultList -> {
+                    queryTokensSetOnce.set(TokenWeightUtil.fetchListOfTokenWeightMap(mapResultList).get(0));
+                    actionListener.onResponse(null);
+                }, actionListener::onFailure)
+            ))
+        );
+        return new SparseEncodingQueryBuilder().fieldName(fieldName)
+            .queryText(queryText)
+            .modelId(modelId)
+            .queryTokensSupplier(queryTokensSetOnce::get);
+    }
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        final MappedFieldType ft = context.fieldMapper(fieldName);
+        validateFieldType(ft);
+
+        Map<String, Float> queryTokens = queryTokensSupplier.get();
+        validateQueryTokens(queryTokens);
+
+        BooleanQuery.Builder builder = new BooleanQuery.Builder();
+        for (Map.Entry<String, Float> entry : queryTokens.entrySet()) {
+            builder.add(FeatureField.newLinearQuery(fieldName, entry.getKey(), entry.getValue()), BooleanClause.Occur.SHOULD);
+        }
+        return builder.build();
+    }
+
+    private static void validateForRewrite(String queryText, String modelId) {
+        if (StringUtils.isBlank(queryText) || StringUtils.isBlank(modelId)) {
+            throw new IllegalArgumentException(
+                QUERY_TEXT_FIELD.getPreferredName() + " and " + MODEL_ID_FIELD.getPreferredName() + " cannot be null."
```

Comment:
please use String.format for this error message

Reply:
Sure, has rewritten all strings having more than 2 args to String.format